File: python/quick-sort.py
#
# @lc app=leetcode id=912 lang=python3
#
# [912] Sort an Array
#

# @lc code=start


class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        self.quicksort(nums, 0, len(nums) - 1)
        return nums

    def quicksort(self, nums, start, end):
        # remember to have exit
        if start >= end:
            return

        left, right = start, end
        # this value should be changing
        pivot = nums[(left + right) // 2]

        # key point 2: every time you compare left & right, it should be
        # left <= right not left < right
        # to skip the pivot, cuz this thing don't need to be sorted any more
        while left <= right:
            while left <= right and nums[left] < pivot:
                left += 1
            while left <= right and nums[right] > pivot:
                right -= 1

            if left <= right:
                nums[left], nums[right] = nums[right], nums[left]
                left += 1
                right -= 1
        #              right  pivot left
        # [9, 5, 2, 9, 15,     20,   70, 40, 80, 55, 75]
        # [9, 5, 2, 9, 15]
        # [70, 40, 80, 55, 75]
        self.quicksort(nums, start, right)
        self.quicksort(nums, left, end)

        # @lc code=end

File: swift/quicksort.swift
class Solution {
    func sortArray(_ nums: [Int]) -> [Int] {
        var nums = nums
        // print(5/1, 5/2, 5/3, 5/4, 5/5)
        quickSort(&nums, 0 , nums.count - 1)
        return nums
    }
    
    func quickSort(_ nums: inout [Int], _ start: Int, _ end: Int) {
        if start >= end {
            return
        }
        
        var l = start, r = end;
        
        let pivot = nums[((start + end) / 2)]
        
        while l <= r {
            while l <= r && nums[l] < pivot { l += 1 }
            while l <= r && nums[r] > pivot { r -= 1 }
            
            if l <= r {
                nums.swapAt(l, r)
                l += 1
                r -= 1
            }
        }
                     
        
        quickSort(&nums, start, r)
        quickSort(&nums, l, end)
    }
}